src/sonic-eventd/src/eventd.cpp

```diff
@@ -0,0 +1,499 @@
+                        ss.str().c_str(), VEC_SIZE(m_events));
+                cap_state = CAP_STATE_LAST;
+                // fall through to save this event in last set.
+            }
+
+        case CAP_STATE_LAST:
+            if (!m_last_events_init) {
+                /* Clear the map, created to ensure memory space available */
+                m_last_events.clear();
+                m_last_events_init = true;
+            }
+            m_last_events[rid] = evt_str;
+            break;
+        }
+    }
+
+out:
+    /*
+     * Capture stop will close the socket which fail the read
+     * and hence bail out.
+     */
+    zmq_close(sock);
+    m_cap_run = false;
+    return;
+}
+
+
+int
+capture_service::set_control(capture_control_t ctrl, event_serialized_lst_t *lst)
+{
+    int ret = -1;
+
+    /* Can go in single step only. */
+    RET_ON_ERR((ctrl - m_ctrl) == 1, "m_ctrl(%d)+1 < ctrl(%d)", m_ctrl, ctrl);
+
+    switch(ctrl) {
+        case INIT_CAPTURE:
+            m_thr = thread(&capture_service::do_capture, this);
+            for(int i=0; !m_cap_run && (i < 100); ++i) {
+                /* Wait max a second for thread to init */
+                this_thread::sleep_for(chrono::milliseconds(10));
+            }
+            RET_ON_ERR(m_cap_run, "Failed to init capture");
+            m_ctrl = ctrl;
+            ret = 0;
+            break;
+
+        case START_CAPTURE:
+            
+            /*
+             * Reserve a MAX_PUBLISHERS_COUNT entries for last events, as we use it only
+             * upon m_events/vector overflow, which might block adding new entries in map
+             * if overall mem consumption is too high. Clearing the map just before use
+             * is likely to help.
+             */
+            for (int i=0; i<MAX_PUBLISHERS_COUNT; ++i) {
+                m_last_events[to_string(i)] = "";
+            }
+            
+            if ((lst != NULL) && (!lst->empty())) {
+                init_capture_cache(*lst);
+            }
+            m_ctrl = ctrl;
+            ret = 0;
+            break;
+
+
+        case STOP_CAPTURE:
+            /*
+             * Caller would have initiated SUBS channel.
+             * Read for CACHE_DRAIN_IN_MILLISECS to drain off cache
+             * before stopping.
+             */
+            this_thread::sleep_for(chrono::milliseconds(CACHE_DRAIN_IN_MILLISECS));
+            stop_capture();
+            ret = 0;
+            break;
+
+        default:
+            SWSS_LOG_ERROR("Unexpected code=%d", ctrl);
+            break;
+    }
+out:
+    return ret;
+}
+
+int
+capture_service::read_cache(event_serialized_lst_t &lst_fifo,
+        last_events_t &lst_last)
+{
+    lst_fifo.swap(m_events);
+    if (m_last_events_init) {
+        lst_last.swap(m_last_events);
+    } else {
+        last_events_t().swap(lst_last);
+    }
+    last_events_t().swap(m_last_events);
+    event_serialized_lst_t().swap(m_events);
```

Comment:
read_cache is internal API that is called right after STOP. Here we swap our cached data with args to avoid copying. If incoming arg had any value, it would be in m_events, so we clear it.